west.yml: Update zcbor to 0.8.0

oyvindronningstad opened the pull request:
- Add support for unordered maps
- Performance improvements
- Naming improvements for generated code
- Bugfixes

oyvindronningstad committed:
west.yml: Update zcbor to 0.8.0 as well as uoscore-uedhoc and MCUboot

Highlights of zcbor 0.8.0:
- Add support for unordered maps
- Performance improvements
- Naming improvements for generated code
- Bugfixes

Update MCUboot and uoscore-uedhoc to bring in changes related to
zcbor 0.8.0.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
lwm2m: Adapt to zcbor 0.8.0

Regenerate from CDDL, update patch, fix references in code.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
mgmt: Adapt to API changes in zcbor 0.8.0

zcbor_new_state and zcbor_tstr_put_term

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
tests: mgmt: Adapt to API changes in zcbor 0.8.0

zcbor_new_state and zcbor_tstr_put_term

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
modules: zcbor: Add config CONFIG_ZCBOR_MAX_STR_LEN

For use with zcbor_tstr_put_term() which needs a maximum string
length.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

SeppoTakalo reviewed (dismissed):
I did not check the code, but at least LwM2M interop tests against Leshan seem to pass. Verified that manually.

So I cannot see anything broken. OK from LwM2M point of view.

oyvindronningstad commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/src/os_mgmt_client.c:
```diff
@@ -138,7 +138,8 @@ int os_mgmt_client_echo(struct os_mgmt_client *client, const char *echo_string)
 	ok = zcbor_map_start_encode(zse, 2) &&
-	     zcbor_tstr_put_lit(zse, "d") && zcbor_tstr_put_term(zse, echo_string) &&
+	     zcbor_tstr_put_lit(zse, "d") &&
+	     zcbor_tstr_put_term(zse, echo_string, CONFIG_ZCBOR_MAX_STR_LEN) &&
```
@nordicjm I agree. I managed to limit it to only be used in tests, and I changed it to be 256 instead of 4096. And I added help text discouraging its use.

nordicjm commented on subsys/stats/stats.c:
```diff
@@ -221,6 +221,10 @@ stats_register(const char *name, struct stats_hdr *hdr)
 	struct stats_hdr *prev;
 	struct stats_hdr *cur;
 
+	if (strnlen(name, CONFIG_STATS_MAX_NAME_LEN) == CONFIG_STATS_MAX_NAME_LEN) {
+		return -ENAMETOOLONG;
+	}
```
Not really sure on this? Seems like an MCUmgr limitation, don't see why it should apply to stats as a whole

oyvindronningstad replied:
It's a sensible check on its own I'd say, but if we don't want to make any changes there, I could make a config in stat_mgmt/Kconfig instead, just for use when encoding.

nordicjm replied:
Stats system has been in zephyr since version 1.11 so would prefer not to make possible breaking changes to it at this point unless they're fixes

nordicjm commented on subsys/mgmt/mcumgr/grp/stat_mgmt/Kconfig:
```diff
@@ -32,6 +32,14 @@ config MCUMGR_GRP_STAT_MAX_NAME_LEN
 	  stat read commands.  If a stat group's name exceeds this limit, it will
 	  be impossible to retrieve its values with a stat show command.
 
+config MCUMGR_STAT_MAX_NAME_LEN
```
Must follow naming rule from top of file:
```
# Options defined in this file should be prefixed:
#  MCUMGR_GRP_STAT_ -- general group options;
```

oyvindronningstad replied:
I changed it to reuse the existing config.

nordicjm commented on west.yml:
```diff
@@ -332,10 +333,11 @@ manifest:
       path: modules/lib/uoscore-uedhoc
     - name: zcbor
-      revision: 67fd8bb88d3136738661fa8bb5f9989103f4599e
+      revision: 0.8.0
```
Need to use a commit ID for this, not a tag

oyvindronningstad replied:
Done

nordicjm reviewed (changes requested):
Message about version update of zcbor should be added into `doc/releases/release-notes-3.6.rst` and details put into `doc/releases/migration-guide-3.6.rst` about what changes applications need to make with the new version

oyvindronningstad committed:
doc: Add release notes and migration guide for zcbor 0.8.0

A short release note referring to the official release notes,
and a migration guide with the most likely changes that need to be made.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
scripts: requirements-extra.txt: Update zcbor from 0.6.0 to 0.8.0

If people are using the zcbor script for code generation, it needs to be
at the latest version

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

de-nordic commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -430,6 +430,32 @@ Networking
+  regenerate the files using zcbor 0.8.0. Note that the names of generated types and members has
+  been overhauled, so the code using the generated code must likely be changed.
+  For example:
+    * Leading single underscores and all double underscores are largely gone,
```
@oyvindronningstad Doc build does not like something in this line
```
/runner/_work/zephyr/zephyr/doc/_build/src/releases/migration-guide-3.6.rst:440: ERROR: Unexpected indentation.
```

kartben commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -430,6 +430,32 @@ Networking
+* If you have zcbor-generated code that relies on the zcbor libraries through Zephyr, you must
+  regenerate the files using zcbor 0.8.0. Note that the names of generated types and members has
+  been overhauled, so the code using the generated code must likely be changed.
+  For example:
+    * Leading single underscores and all double underscores are largely gone,
+    * Names sometimes gain suffixes like ``_m`` or ``_l`` for disambiguation.
+    * All enum (choice) names have now gained a ``_c`` suffix, so the enum name no longer matches
+      the corresponding member name exactly (because this broke C++ namespace rules).
```
probably should be: 

```suggestion
  For example:
  
  * Leading single underscores and all double underscores are largely gone,
  * Names sometimes gain suffixes like ``_m`` or ``_l`` for disambiguation.
  * All enum (choice) names have now gained a ``_c`` suffix, so the enum name no longer matches
      the corresponding member name exactly (because this broke C++ namespace rules).
```

oyvindronningstad commented:
Thanks @de-nordic and @kartben, doc build fixed.

nordicjm commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -430,6 +430,33 @@ Networking
+  been overhauled, so the code using the generated code must likely be changed.
+  For example:
+
+  * Leading single underscores and all double underscores are largely gone,
```
(only if another change is requested) replace comma with full stop
